[css-anchor-position] Add flip-x/flip-y to position-try-fallback

nt1m opened the pull request:
#### 335c6246019b91132d44a06a0ecb3a59bc694591
<pre>
[css-anchor-position] Add flip-x/flip-y to position-try-fallback
<a href="https://bugs.webkit.org/show_bug.cgi?id=301366">https://bugs.webkit.org/show_bug.cgi?id=301366</a>
<a href="https://rdar.apple.com/163282036">rdar://163282036</a>

Reviewed by Elika Etemad.

CSSWG resolution: <a href="https://github.com/w3c/csswg-drafts/issues/12869#issuecomment-3406978831">https://github.com/w3c/csswg-drafts/issues/12869#issuecomment-3406978831</a>

Tests from: <a href="https://github.com/web-platform-tests/wpt/pull/55616">https://github.com/web-platform-tests/wpt/pull/55616</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-alignment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-alignment.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-wm-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-wm.html:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PositionTry.cpp:
(WebCore::CSSPropertyParserHelpers::consumePositionTryFallbacks):
* Source/WebCore/rendering/style/PositionTryFallback.cpp:
(WebCore::Style::operator&lt;&lt;):
* Source/WebCore/rendering/style/PositionTryFallback.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::applyTryTacticsToInset):
(WebCore::Style::AnchorPositionEvaluator::resolvePositionTryFallbackProperty):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData):
(WebCore::Style::flipPositionAreaByPhysicalAxis):
(WebCore::Style::BuilderConverter::convertPositionArea):

Canonical link: <a href="https://commits.webkit.org/302057@main">https://commits.webkit.org/302057@main</a>
</pre>

fantasai commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -644,6 +644,18 @@ static std::pair<CSSPropertyID, bool> applyTryTacticsToInset(CSSPropertyID prope
+                propertyID = getOppositeInset(propertyID);
+                isFlipped = true;
+            }
+            break;
```
Nit: I would move these up so that they're next to FlipInline and FlipBlock, which have parallel structure.

fantasai commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -554,13 +554,23 @@ inline StyleSelfAlignmentData BuilderConverter::convertSelfOrDefaultAlignmentDat
             switch (tactic) {
+            case PositionTryFallback::Tactic::FlipX:
+                if ((builderState.cssPropertyID() == CSSPropertyAlignSelf && writingMode.isHorizontal())
+                    || (builderState.cssPropertyID() == CSSPropertyJustifySelf && !writingMode.isHorizontal()))
```
I think you can just write 
```suggestion
                if (writingMode.isHorizontal() ? (builderState.cssPropertyID() == CSSPropertyJustifySelf) : (builderState.cssPropertyID() == CSSPropertyJustifySelf))
```

Also I think your logic is backwards, horizontal flip-x should be flipping justifySelf, right?

fantasai reviewed (changes requested):
Also this needs tests.

nt1m committed:
[css-anchor-position] Add flip-x/flip-y to position-try-fallback
https://bugs.webkit.org/show_bug.cgi?id=301366
rdar://163282036

Reviewed by Elika Etemad.

CSSWG resolution: https://github.com/w3c/csswg-drafts/issues/12869#issuecomment-3406978831

Tests from: https://github.com/web-platform-tests/wpt/pull/55616

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-fallbacks-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-try-parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-alignment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-alignment.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-wm-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-wm.html:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PositionTry.cpp:
(WebCore::CSSPropertyParserHelpers::consumePositionTryFallbacks):
* Source/WebCore/rendering/style/PositionTryFallback.cpp:
(WebCore::Style::operator<<):
* Source/WebCore/rendering/style/PositionTryFallback.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::applyTryTacticsToInset):
(WebCore::Style::AnchorPositionEvaluator::resolvePositionTryFallbackProperty):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData):
(WebCore::Style::flipPositionAreaByPhysicalAxis):
(WebCore::Style::BuilderConverter::convertPositionArea):

Canonical link: https://commits.webkit.org/302057@main
